### PR TITLE
Parallel evaluation of tasks

### PR DESCRIPTION
**Pull request by Neeratyoy**
#### What does this PR implement/fix? Explain your changes.
Uses `joblib` to evaluate different folds of a task in parallel, in order to speed-up the purely sequential evaluations of tasks right now.

#### How should this PR be tested?
The unit tests added to test the accuracy of the distribution of evaluation and collation of results.

### CODE REVIEW

File: openml/runs/functions.py [hunk 1]
```diff
@@ -54,6 +55,7 @@ def run_model_on_task(
     upload_flow: bool = False,
     return_flow: bool = False,
     dataset_format: str = "dataframe",
+    n_jobs: int = None,
```
Comment (mfeurer) [inline openml/runs/functions.py:58]: Could you please explicitly write `n_jobs: Optional[int] = None`?

Comment (mfeurer) [reply openml/runs/functions.py:58]: Also, please add it to the docstring

File: openml/runs/functions.py [hunk 2]
```diff
@@ -146,6 +149,7 @@ def run_flow_on_task(
     add_local_measures: bool = True,
     upload_flow: bool = False,
     dataset_format: str = "dataframe",
+    n_jobs: int = None,
```
Comment (mfeurer) [inline openml/runs/functions.py:152]: Same as above.

File: openml/runs/functions.py [hunk 3]
```diff
@@ -615,6 +603,75 @@ def _calculate_local_measure(sklearn_fn, openml_name):
...
+        raise NotImplementedError(task.task_type)
+    config.logger.info(
+        "Going to execute flow '%s' on task %d for repeat %d fold %d sample %d.",
+        flow.name,
+        task.task_id,
+        rep_no,
+        fold_no,
+        sample_no,
+    )
+    (pred_y, proba_y, user_defined_measures_fold, trace,) = extension._run_model_on_fold(
```
Comment (mfeurer) [inline openml/runs/functions.py:663]: Could you please remove these brackets?

File: tests/test_openml/test_config.py [hunk 4]
```diff
@@ -11,6 +11,32 @@ def test_config_loading(self):
...
+    def test_get_config_as_dict(self):
+        """ Checks if the current configuration is returned accurately as a dict. """
+        config = openml.config.get_config_as_dict()
+        self.assertIsInstance(config, dict)
+        self.assertEqual(len(config), 6)
+        self.assertEqual(config.get("server"), "https://test.openml.org/api/v1/xml")
+        self.assertEqual(config.get("apikey"), "610344db6388d9ba34f6db45a3cf71de")
+        self.assertEqual(config.get("cachedir"), self.workdir)
+        self.assertEqual(config.get("avoid_duplicate_runs"), False)
+        self.assertEqual(config.get("max_retries"), 20)
```
Comment (mfeurer) [inline tests/test_openml/test_config.py:23]: I believe there's one test missing? However, why don't you test for dict equality as in #880?

File: tests/test_runs/test_run_functions.py [hunk 5]
```diff
@@ -1575,3 +1576,86 @@ def test_format_prediction_task_regression(self):
...
+                task=task,
+                add_local_measures=True,
+                dataset_format="array",  # "dataframe" would require handling of categoricals
+                n_jobs=self.n_jobs,
+            )
+        self.assertEqual(type(res[0]), list)
+        self.assertEqual(len(res[0]), num_instances)
+        self.assertEqual(len(res[0][0]), line_length)
+        self.assertEqual(len(res[2]), 7)
+        self.assertEqual(len(res[3]), 7)
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1603]: As checking the call count is no longer necessary, could you please make these checks stricter, for example by checking for exact predictions?

File: tests/test_runs/test_run_functions.py [hunk 6]
```diff
@@ -1575,3 +1576,86 @@ def test_format_prediction_task_regression(self):
...
+                res = openml.runs.functions._run_task_get_arffcontent(
+                    flow=flow,
+                    extension=self.extension,
+                    model=clf,
+                    task=task,
+                    add_local_measures=True,
+                    dataset_format="array",  # "dataframe" would require handling of categoricals
+                    n_jobs=n_jobs,
+                )
+            self.assertEqual(type(res[0]), list)
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1653]: Please use isinstance here

File: tests/test_runs/test_run_functions.py [hunk 7]
```diff
@@ -1575,3 +1576,86 @@ def test_format_prediction_task_regression(self):
...
+                    "criterion": ["gini", "entropy"],
+                },
+                random_state=1,
+                cv=sklearn.model_selection.StratifiedKFold(
+                    n_splits=2, shuffle=True, random_state=1
+                ),
+                n_iter=5,
+                n_jobs=n_jobs,
+            )
+            with parallel_backend(backend, n_jobs=n_jobs):
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1643]: I'm wondering whether we can create a mock which checks for the inner function being executed in parallel? If the call_count is larger than zero, then it wasn't?

File: tests/test_runs/test_run_functions.py (hunk 5 above)
Comment (Neeratyoy) [reply tests/test_runs/test_run_functions.py:1604]: Added a check for `predictive_accuracy` per fold.

File: tests/test_runs/test_run_functions.py (hunk 7 above)
Comment (Neeratyoy) [reply tests/test_runs/test_run_functions.py:1659]: Added a mock as you mentioned.

Comment (mfeurer) [reply tests/test_runs/test_run_functions.py:1663]: Hey, I think you added this only to the other unit test, not this one. I think it's a great addition to the other unit test, but could you please also add it here?

Comment (mfeurer) [reply tests/test_runs/test_run_functions.py:1663]: Also, looking at this again, I'm not really sure what that mock checks for?

File: openml/testing.py [hunk 8]
```diff
@@ -110,6 +110,8 @@ def setUp(self, n_levels: int = 1):
         # Increase the number of retries to avoid spurious server failures
         self.connection_n_retries = openml.config.connection_n_retries
         openml.config.connection_n_retries = 10
+        # Number of processes to parallelize any evaluation made by a unit test
+        self.n_jobs = 2
```
Comment (Neeratyoy) [inline openml/testing.py:114]: No need of a class-variable!

File: tests/test_runs/test_run_functions.py (hunk 7 above)
Comment (Neeratyoy) [reply tests/test_runs/test_run_functions.py:1669]: Tried fixing it with a different function to mock.

**Review by mfeurer — APPROVED** [review]

**Review by mfeurer — APPROVED** [review]
